### PR TITLE
[12.0] add account_asset_management reporting

### PR DESCRIPTION
**Pull request by luc-demeyer**
enhancements compared to 11.0 account_asset_management_xls (https://github.com/OCA/account-financial-tools/pull/701):

- period based reporting (compared to Fiscal Year based in previous versions)
- multiple reporting structures thanks to m2m nature of asset groups
- draft assets can be added to the report
- inconsistent reporting structures are detected

### CODE REVIEW

**Comment by ivantodorovich** [comment]
Hi @luc-demeyer  :)
thanks for this :)

can you please check travis errors?
there are a few LINT errors and also please add reporting-engine to oca_dependencies so we avoid the missing depenency error

**Comment by ivantodorovich** [comment]
Awesome @luc-demeyer ! :)
Now that runbot is building:

**Functional review:**

1. Asset group

Assets Group is required on the wizard, but it's not required on the asset itself. So it's impossible to report assets that has no group.

Maybe the asset group filter should be optional, don't you think?

image
image

---

2. The company_id field should be shown only if the user has the multi-company group.

image

---

3. Report menu location

I think it makes more sense to move it to _Reporting > Assets > Financial Report_
The trend in odoo is to group all the reports in one menu

image

**Comment by ivantodorovich** [comment]
Also this would be great: https://github.com/OCA/account-financial-tools/issues/878

**Comment by ivantodorovich** [comment]
Hi @luc-demeyer 

Do you plan on working on this?
We'd like to work on this to add more information to the report.

We are more than happy to pick up where you left, if you don't have the time. Just let me know :)

**Comment by luc-demeyer** [comment]
@ivantodorovich 
I intend to rebase the PR after merge of https://github.com/OCA/account-financial-tools/pull/925
and add a couple of small changes. If you can review PR925 to speed up the merge that would help.

If you have time you can of course always make a PR on top of my PR so that I can review and add it all together. I am interested to know what are the extra informations you intend to add.

**Comment by ivantodorovich** [comment]
This is an example report (without sensible data) we're using in a 9.0 project.

[Account.Asset.Report.xlsx](https://github.com/OCA/account-financial-tools/files/4147747/Account.Asset.Report.xlsx)

The original report has 300+ assets.

We'd like to integrate it with this PR (or another PR, after this one is merged).

Some features
- Group by category, with subtotals
- Previous depreciation (depreciation until report start date)
- Period depreciation (depreciation during report start/end date)
- Accumulated depreciation 
- Clean look

**Comment by luc-demeyer** [comment]
Looking to the information :

 

That’s already present in the report.

As soon as you define your reporting tag structure you get subtotals per reporting tag.

 

Luc

**Comment by ivantodorovich** [comment]
> As soon as you define your reporting tag structure you get subtotals per reporting tag.

What do you mean? :thinking: 

I feel like I'm missing something huge here :sweat_smile: 

Where are we supposed to configure these tags?


This is the report I get from this module:
[Testinggroup_asset_report.xlsx](https://github.com/OCA/account-financial-tools/files/4148024/Testinggroup_asset_report.xlsx)

**Comment by luc-demeyer** [comment]
The tags replace the "parent_id" reporting structure (has been there since version 6 of this module) but this concept is richer since now you can overlay multiple reporting structures (e.g. a legal one and one for your specific business needs). 
The account_asset_management_xls reporting module (which will now be integrated in the base module with this PR) has always given  subtotals.

Concerning reporting structures, have a look at https://github.com/luc-demeyer/noviat-apps/tree/11.0/l10n_be_asset_management_data.
This one adds the legal belgian asset reporting structure. As soon as the new tag based reporting structure is merged I'll publish an update of this localisation module for v12.

**Comment by ivantodorovich** [comment]
So, what you mean is that we should create a `l10n_fr_asset_management_data` module, to implement this structure?

And, basically, extend `['account.asset']`: to overload `_xls_acquisition_template`, and other `_xls_*` methods, :thinking: ?

**Comment by luc-demeyer** [comment]
@ivantodorovich 
in case there exist a legal french reporting structure for assets than it is indeed recommended to create a small data module to add this structure.
If you want to add extra fields to the standard report than I propose that we extend the _xls_acquisition_template in the base module because I can hardly imagine that your requirement will be limited to France only.
In case the requirement is generic than it would also be best to add the field also to the _xls_acquisition_fields. 
Localisation modules can freely add/remove fields by modifying the _xls_acquisition_fields. That's a five minutes effort.

**Comment by luc-demeyer** [comment]
@ivantodorovich 
I have rebased this PR and pushed an update  PR which implements the suggestions made by you in your review:
- Asset group is now optional in the reporting wizard. A flat report is generated when no group is 
 specified (although I still recommend to add localisation modules with asset reporting structure, cf. https://github.com/Noviat/noviat-apps/tree/12.0/l10n_be_asset_management_data for an example.
- the asset report is now moved under the "reporting" main menu as suggested by you
- added multi-company group on company-id

On top of this I also implemented a couple of small usability enhancements.
Can you review ?

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+indent_size = 2
+
+# Do not configure editor for libs and autogenerated content
+[*/static/{lib,src/lib}/**,*/static/description/index.html,*/readme/../README.rst]
+charset = unset
+end_of_line = unset
+indent_size = unset
+indent_style = unset
+insert_final_newline = false
+trim_trailing_whitespace = false
```
Comment (ivantodorovich) [inline .editorconfig:20]: Is this file supposed to be here?

**Review by ivantodorovich — APPROVED** [review]
:+1: looks good! thanks!

**Comment by luc-demeyer** [comment]
This one is ready for merge (and port to 13.0)

@pedrobaeza 
Can you merge (ocabot major) and also answer ivantodorovich remarks on the .editorconfig ?
I'll make the 13.0 PR after merge.

**Comment by pedrobaeza** [comment]
@carlosdauden can you please review this?

@luc-demeyer is it possible to squash initial bunch of commits from @astirpe ?

File: account_asset_management/models/account_asset_group.py [hunk 2]
```diff
@@ -23,7 +24,42 @@ class AccountAssetGroup(models.Model):
...
+    @api.multi
+    def name_get(self):
+        result = []
+        params = self.env.context.get('params')
+        list_view = params and params.get('view_type') == 'list'
+        for rec in self:
+            if list_view:
+                name = rec.code
+            else:
+                name = rec.code + ' ' + rec.name
```
Comment (qgroulard) [inline account_asset_management/models/account_asset_group.py]: `code` might be empty and cause an error.

**Comment by luc-demeyer** [comment]
@qgroulard 
Thanks for your feedback.
Fixed via https://github.com/OCA/account-financial-tools/pull/892/commits/5920f1fd7199d408c95b94e0f06363e4d1c58364

I now truncate the name when there is no code.
The asset list view now gets the following look & feel:

image

**Comment by luc-demeyer** [comment]
@pedrobaeza 
I don't feel very comfortable to squash a bunch of commits from 2018. 
Can you give me the syntax on how to do it to avoid that I squash too much ?

**Comment by pedrobaeza** [comment]
I can do it if you prefer.

**Comment by luc-demeyer** [comment]
@pedrobaeza 
as you can imagine I prefer that you squash the historical commits.

**Comment by pedrobaeza** [comment]
Hehe, OK, it was just a formality for not doing it unadvertised. I'll do it soon and tell you back. I also pinged my colleague for reviewing this and finally merge it.

**Review by carlosdauden — APPROVED** [review]
Good work

**Comment by luc-demeyer** [comment]
@pedrobaeza can you merge ?

**Comment by pedrobaeza** [comment]
Fixed the mess in #980, as I can't push to your branch. Please never ever use GitHub conflict resolution nor the equivalent in git through merge operation. It's not first time I have to fix similar things and it requires a lot of time to fix conflicts. You should check the method you use. If you tell me the git used commands, I can tell you where is the problem.

Any way, closing this and continuing in the other, where my main concern is the added dependency.

**Pull request closed without merging**